### ISSUE TITLE
StoreUmbracoFormsInDb setting causes: Invalid object name 'UFUserStartFolders'

### ISSUE THREAD

**Issue by shearer3000** [umbraco/Umbraco.Forms.Issues#888]
Hello

I am upgrading an Umbraco 7 site that has Umbraco Forms 7.2.1 installed.

I started by upgrading to the latest version of umbraco 7, and latest Umbraco Forms 7.5.9

Then, I upgrade the database to latest Umbraco 8 (8.18.5) and UF 8.0.0, and then also onto the latest Umbraco forms 8 version, 8.13.2

At which point I also changed the “StoreUmbracoFormsInDb” to true

When I run the site at this stage and go to the UF section I get:
`Invalid object name 'UFUserStartFolders'`

I can see that table (and other tables related to storing UF in the database) don’t exist.

How do I trigger the migration that creates them?

Ultimately I'm upgrading the Umbraco db to 10.2, and StoreUmbracoFormsInDb=true is a prerequisite of UF10.x.
Currently, the umbraco 10 site isn't starting with the following boot error which i understand from googling is UF-related:
` Umbraco.Cms.Core.Exceptions.BootFailedException: The migration plan does not support migrating from state "6e692c5d-c670-4c34-af17-28d8dbf0dcd2". -> System.InvalidOperationException: The migration plan does not support migrating from state "6e692c5d-c670-4c34-af17-28d8dbf0dcd2"`

Many thanks
Andrew

**Comment by AndyButland** [comment]
Hi @shearer3000 - yes, unfortunately it looks like we missed a migration case here, where someone is on the last version of Forms 8, and then switches into the "forms in database" mode.  The migration plan isn't picking up that change and creating the necessary tables.

I think we can patch fix this so will look to do so for the next update.

In the meantime if you can follow this pattern, you should be OK:

- When you are on Umbraco CMS V8 and Forms 8.0.0, install first Forms 8.7 (I think actually any version from 8.5 to 8.12 should be OK, as it's only the last one where we've missed handling this migration path).
- Switch then into `StoreUmbracoFormsInDb` = `true`.
- Check your forms have migrated and are visible in the back-office.
- Then upgrade to Forms 8.13 and carry on from there.

**Comment by AndyButland** [comment]
We'll include an update in the next patch that will handle this migration without the workaround described above.

**Comment by shearer3000** [comment]
Great - thanks for looking into this and providing some interim steps. will try those now :)

**Comment by shearer3000** [comment]
hi Andy - i tried the steps of Umbraco Forms 8.0.0 to 8.7.0 but after setting "StoreUmbracoFormsInDb" to true and running the site I get:

`System.NullReferenceException: Object reference not set to an instance of an object.
  at Umbraco.Forms.Core.Migrations.V_8_5_0.FormFilesAndWorkflowFilesToDb.<>c__DisplayClass6_0.<Migrate>b__0(IWorkflow workflow)
  at System.Linq.Enumerable.WhereListIterator`1.MoveNext()
  at Umbraco.Forms.Core.Migrations.V_8_5_0.FormFilesAndWorkflowFilesToDb.Migrate()
  at Umbraco.Forms.Core.Migrations.V_8_5_0.AddFormsDbStorage.Migrate()
  at Umbraco.Core.Migrations.MigrationBase.Umbraco.Core.Migrations.IMigration.Migrate() in D:\a\1\s\src\Umbraco.Core\Migrations\MigrationBase.cs:line 73
  at Umbraco.Core.Migrations.MigrationPlan.Execute(IScope scope, String fromState, IMigrationBuilder migrationBuilder, ILogger logger) in D:\a\1\s\src\Umbraco.Core\Migrations\MigrationPlan.cs:line 313
  at Umbraco.Core.Migrations.Upgrade.Upgrader.Execute(IScopeProvider scopeProvider, IMigrationBuilder migrationBuilder, IKeyValueService keyValueService, ILogger logger) in D:\a\1\s\src\Umbraco.Core\Migrations\Upgrade\Upgrader.cs:line 67
  at Umbraco.Forms.Core.Components.UmbracoFormsComponent.ExecuteDatabaseUpgrades()
  at Umbraco.Forms.Core.Components.UmbracoFormsComponent.Initialize()
  at Umbraco.Core.Composing.ComponentCollection.Initialize() in D:\a\1\s\src\Umbraco.Core\Composing\ComponentCollection.cs:line 32
  at Umbraco.Core.Runtime.CoreRuntime.Boot(IRegister register, DisposableTimer timer) in D:\a\1\s\src\Umbraco.Core\Runtime\CoreRuntime.cs:line 219`

any clues what line 313 is trying to do?
thanks

**Comment by AndyButland** [comment]
Sorry @shearer3000 - I've had a look back at that version, and I can't see from scanning the code why you should be getting a null reference there.

In this section of code the migration is looking to go through the workflow definition files stored on disk and add them to the database.

Would it be possible to share with me the copies of the form definition files you have stored on disk (i.e. the contents of `App_Data\UmbracoForms\Data`, and I can try to replicate/debug?  abl AT umbraco DOT dk if so.

**Comment by AndyButland** [comment]
Hi @shearer3000 - thanks for sending over the details.  I ran it on my local setup and could replicate the problem.  When I stepped through I could see there was one corrupted file in there - `workflows/92b0426c-b03b-4246-9634-6f72b4bbc6ef.json`.  If you open that up you'll see it's not got valid contents in it.

So if you:

- Remove that file.
- Reset the migration state to `6822cdca-1dc4-4983-b189-9c45314e8747`, via this SQL:

```
update umbracoKeyValue
set value = '6e692c5d-c670-4c34-af17-28d8dbf0dcd2'
where [key] = 'Umbraco.Core.Upgrader.State+UmbracoForms'
```
- Make sure you have `<setting key="StoreUmbracoFormsInDb" value="True" />`
- Restart the web application

It should run to completion then with the form definitions loaded into the database.

I'll add an update to the upcoming patch to ensure we detect a deserialization failure like this, log a warning and move on, rather than throwing an exception.
